README.md
# CG-3 IDE

* https://visl.sdu.dk/cg3.html
* https://groups.google.com/group/constraint-grammar

Currently licensed under GNU General Public License version 3 or newer.
See CONTRIBUTING.md for the copyright assignment agreement.


ChangeLog
2014-01-29  tino

	* [r9683] CMakeLists.txt: PPAs can now be built for Ubuntu 14.04

2014-01-28  tino

	* [r9669] CMakeLists.txt: Ubuntu 13.04 is EOL, so drop from PPA

2013-12-21  tino

	* [r9508] ChangeLog, scripts/CG3-IDE.wxs, scripts/pkg-windows.sh,
	  src/version.hpp: Qt 5.2.0

2013-12-01  tino

	* [r9455] CMakeLists.txt, ChangeLog, src/version.hpp,
	  update-revision.pl: Use svn2cl to generate the ChangeLog

2013-11-25  tino

	* [r9407] CMakeLists.txt, IDE.pro, TODO, src/GrammarEditor.cpp,
	  src/GrammarHighlighter.cpp, src/GrammarState.hpp,
	  src/version.hpp, todo.sh, update-revision.pl: Added explicit
	  version number. Upped parse recursion limit to 10000 (from 1000).
	  COPY EXCEPT parses.

2013-11-03  tino

	* [r9323] CMakeLists.txt: Revision bump
	* [r9322] scripts/CG3-IDE.wxs, scripts/pkg-windows.sh,
	  src/GrammarEditor.cpp, src/GrammarEditor.hpp,
	  src/GrammarEditor.ui, src/GrammarHighlighter.cpp: Toggle removed
	  readings

2013-10-17  tino

	* [r9293] CMakeLists.txt: Add Saucy to PPA

2013-09-12  tino

	* [r9220] src/OptionsDialog.cpp: OS X weirdness workaround

2013-06-25  tino

	* [r9082] cg3ide.iconset/icon-fullsize.psd,
	  cg3ide.iconset/icon_128x128.png, cg3ide.iconset/icon_16x16.png,
	  cg3ide.iconset/icon_256x256.png, cg3ide.iconset/icon_32x32.png:
	  -x

2013-06-20  tino

	* [r9067] src/GrammarEditor.cpp, src/GrammarEditor.hpp,
	  src/GrammarEditor.ui, src/GrammarHighlighter.cpp,
	  src/GrammarHighlighter.hpp: Added jump to sections
	* [r9066] IDE.pro, Processor.pro, cg3ide.pro, cg3ide.rc,
	  scripts/pkg-windows.sh, src/GotoLine.cpp, src/GotoLine.hpp,
	  src/GotoLine.ui, src/GrammarEditor.cpp, src/GrammarEditor.hpp,
	  src/GrammarEditor.ui, src/GrammarHighlighter.cpp,
	  src/GrammarHighlighter.hpp, src/GrammarState.hpp,
	  src/OptionsDialog.cpp, src/OptionsDialog.hpp,
	  src/OptionsDialog.ui, src/Processor.cpp, src/Processor.hpp,
	  src/Processor.ui, src/StreamHighlighter.cpp,
	  src/StreamHighlighter.hpp, src/inlines.hpp, src/main.cpp,
	  src/types.hpp: Add svn:eol-style native for any file Qt Creator
	  is expected to touch.

2013-06-19  tino

	* [r9061] IDE.pro, cg3ide.icns, cg3ide.ico, cg3ide.iconset,
	  cg3ide.iconset/icon-fullsize.psd,
	  cg3ide.iconset/icon_128x128.png, cg3ide.iconset/icon_16x16.png,
	  cg3ide.iconset/icon_256x256.png, cg3ide.iconset/icon_32x32.png,
	  cg3ide.rc: Icons
	* [r9058] CMakeLists.txt: Strip binaries; Upload to /cg3

2013-06-18  tino

	* [r9057] CMake, CMake/FindBoost.cmake, CMake/FindICU.cmake,
	  CMakeLists.txt, src/CMakeLists.txt: CPack .deb helpers
	* [r9056] AUTHORS, COPYING, LICENSE, src/GotoLine.cpp,
	  src/GotoLine.hpp, src/GrammarEditor.cpp, src/GrammarEditor.hpp,
	  src/GrammarHighlighter.cpp, src/GrammarHighlighter.hpp,
	  src/GrammarState.hpp, src/OptionsDialog.cpp,
	  src/OptionsDialog.hpp, src/Processor.cpp, src/Processor.hpp,
	  src/StreamHighlighter.cpp, src/StreamHighlighter.hpp,
	  src/inlines.hpp, src/main.cpp, src/types.hpp: Added copyright and
	  GPLv3 boilerplate
	* [r9055] scripts/pkg-osx.sh: Packaging
	* [r9054] scripts/pkg-osx.sh: Packaging
	* [r9052] scripts/CG3-IDE.wxs, scripts/osx-deploy-etc.sh,
	  scripts/pkg-osx.sh, scripts/pkg-windows.sh, src/inlines.hpp:
	  Packaging

2013-06-17  tino

	* [r9050] src/GrammarEditor.cpp, src/GrammarHighlighter.cpp,
	  src/GrammarState.hpp: Warn on "foo "bar"
	* [r9049] COLA.pro, cg3ide.pro, scripts/CG3-IDE.wxs, specs: CG-3
	  IDE
	* [r9048] .:



src/GrammarState.hpp
/*
* Copyright (C) 2013, GrammarSoft ApS
* Developed by Tino Didriksen <mail@tinodidriksen.com> for GrammarSoft ApS (http://grammarsoft.com/)
* Development funded by Tony Berber Sardinha (http://www2.lael.pucsp.br/~tony/), São Paulo Catholic University (http://pucsp.br/), CEPRIL (http://www2.lael.pucsp.br/corpora/), CNPq (http://cnpq.br/), FAPESP (http://fapesp.br/)
*
* This file is part of CG-3 IDE
*
* CG-3 IDE is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* CG-3 IDE is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with CG-3 IDE.  If not, see <http://www.gnu.org/licenses/>.
*/

#pragma once
#ifndef GRAMMARSTATE_HPP_cc7194f1bd3a13d1dca4d5a1c31f83d81877a7f7
#define GRAMMARSTATE_HPP_cc7194f1bd3a13d1dca4d5a1c31f83d81877a7f7

#include <QtWidgets>
#include <stdint.h>

enum enum_state {
    S_NONE           =         0,
    S_TAGLIST        = (1 <<  0),
    S_TAGLIST_INLINE = (1 <<  1),
    S_COMPOSITETAG   = (1 <<  2),
    S_EQUALS         = (1 <<  3),
    S_SEMICOLON      = (1 <<  4),
    S_COMMENT        = (1 <<  5),
    S_TAG            = (1 <<  6),
    S_SETNAME        = (1 <<  7),
    S_SETOP          = (1 <<  8),
    S_SET_INLINE     = (1 <<  9),
    S_CONTEXT        = (1 << 10),
    S_RULE           = (1 << 11),
    S_TMPLNAME       = (1 << 12),
    S_ERROR          = (1 << 13),
    S_RULE_FLAG      = (1 << 14),
    S_CONTEXT_POS    = (1 << 15),
    S_CONTEXT_TMPL   = (1 << 16),
    S_CONTEXT_OP     = (1 << 17),
    S_PAR_START      = (1 << 18),
    S_PAR_STOP       = (1 << 19),
    S_BARRIER        = (1 << 20),
    S_LINK           = (1 << 21),
    S_SQBRACKET_STOP = (1 << 22),
    S_TARGET         = (1 << 23),
    S_IF             = (1 << 24),
    S_CONTEXT_LIST   = (1 << 25),
    S_TO_FROM        = (1 << 26),
    S_BEFORE_AFTER   = (1 << 27),
    S_WITH           = (1 << 28),
    S_WITHCHILD      = (1 << 29),
    S_ONCE_ALWAYS    = (1 << 30),
    S_EXCEPT         = (1 << 31)
};
Q_DECLARE_FLAGS(State, enum_state)
Q_FLAGS(State)
Q_DECLARE_OPERATORS_FOR_FLAGS(State)

inline QString StateText(State state) {
    QString rv;
    if (state & S_NONE) {
        rv.append("|S_NONE");
    }
    if (state & S_TAGLIST) {
        rv.append("|S_TAGLIST");
    }
    if (state & S_TAGLIST_INLINE) {
        rv.append("|S_TAGLIST_INLINE");
    }
    if (state & S_COMPOSITETAG) {
        rv.append("|S_COMPOSITETAG");
    }
    if (state & S_EQUALS) {
        rv.append("|S_EQUALS");
    }
    if (state & S_SEMICOLON) {
        rv.append("|S_SEMICOLON");
    }
    if (state & S_COMMENT) {
        rv.append("|S_COMMENT");
    }
    if (state & S_TAG) {
        rv.append("|S_TAG");
    }
    if (state & S_SETNAME) {
        rv.append("|S_SETNAME");
    }
    if (state & S_SETOP) {
        rv.append("|S_SETOP");
    }
    if (state & S_SET_INLINE) {
        rv.append("|S_SET_INLINE");
    }
    if (state & S_CONTEXT) {
        rv.append("|S_CONTEXT");
    }
    if (state & S_RULE) {
        rv.append("|S_RULE");
    }
    if (state & S_TMPLNAME) {
        rv.append("|S_TMPLNAME");
    }
    if (state & S_ERROR) {
        rv.append("|S_ERROR");
    }
    if (state & S_CONTEXT_POS) {
        rv.append("|S_CONTEXT_POS");
    }
    if (state & S_CONTEXT_TMPL) {
        rv.append("|S_CONTEXT_TMPL");
    }
    if (state & S_CONTEXT_OP) {
        rv.append("|S_CONTEXT_OP");
    }
    if (state & S_PAR_START) {
        rv.append("|S_PAR_START");
    }
    if (state & S_PAR_STOP) {
        rv.append("|S_PAR_STOP");
    }
    if (state & S_BARRIER) {
        rv.append("|S_BARRIER");
    }
    if (state & S_LINK) {
        rv.append("|S_LINK");
    }
    if (state & S_SQBRACKET_STOP) {
        rv.append("|S_SQBRACKET_STOP");
    }
    if (state & S_TARGET) {
        rv.append("|S_TARGET");
    }
    if (state & S_IF) {
        rv.append("|S_IF");
    }
    if (state & S_CONTEXT_LIST) {
        rv.append("|S_CONTEXT_LIST");
    }
    if (state & S_TO_FROM) {
        rv.append("|S_TO_FROM");
    }
    if (state & S_BEFORE_AFTER) {
        rv.append("|S_BEFORE_AFTER");
    }
    if (state & S_WITH) {
        rv.append("|S_WITH");
    }
    if (state & S_WITHCHILD) {
        rv.append("|S_WITHCHILD");
    }
    if (state & S_ONCE_ALWAYS) {
        rv.append("|S_ONCE_ALWAYS");
    }
    if (state & S_RULE_FLAG) {
        rv.append("|S_RULE_FLAG");
    }
    return rv.trimmed().replace(QRegExp("^\\|"), "");
}

class GrammarState : public QTextBlockUserData {
public:
    GrammarState() {
    }

    GrammarState(const QVector<State>& stack) : stack(stack) {
    }

    QVector<State> stack;
    QString error;

    typedef QMap<int,State> tokens_t;
    tokens_t tokens;

    typedef QMultiMap<QPair<int,int>,QString> warnings_t;
    warnings_t warnings;
};

#endif // GRAMMARSTATE_HPP_cc7194f1bd3a13d1dca4d5a1c31f83d81877a7f7
